### PR TITLE
Bug/show user information

### PR DESCRIPTION
**Pull request by SirWayne**
1. Create a formatter component, which formats the user details.
2. Add table cell tooltip for user information
3. Add label tooltip for user information

Signed-off-by: SirWayne <dennis.melzer@bosch-si.com>

### CODE REVIEW

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+     * characters.
+     * 
+     * @see {@link UserDetailsFormatter#loadAndFormatUsername(String, int)}
+     * 
+     * @param username
+     *            the user name
+     * @return the formatted user name (max 100 characters) cannot be <null>
+     */
+    public static String loadAndFormatUsername(final String username) {
+        return loadAndFormatUsername(username, 50);
```
Comment (schabdo) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java]: Javadoc promised that I'll get up to 100 character

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            return userDetails.getUsername();
+        }
+
+        final UserPrincipal userPrincipal = (UserPrincipal) userDetails;
+
+        final String trimmedFirstname = trimAndFormatDetail(userPrincipal.getFirstname(), expectedNameLength);
+        final String trimmedLastname = trimAndFormatDetail(userPrincipal.getLastname(), expectedNameLength);
+
+        if (StringUtils.isEmpty(trimmedFirstname) && StringUtils.isEmpty(trimmedLastname)) {
+            return StringUtils.substring(userPrincipal.getLoginname(), 0, 2 * expectedNameLength);
```
Comment (schabdo) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java]: Why you multiply the length with 2? Can't you pass the original expected length in here? That's somehow weird.

Comment (schabdo) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java]: As you can see in line 48. I am already confused by the parameter

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+    private static UserDetails getCurrentUser() {
+        final SecurityContext context = (SecurityContext) VaadinService.getCurrentRequest().getWrappedSession()
+                .getAttribute(HttpSessionSecurityContextRepository.SPRING_SECURITY_CONTEXT_KEY);
+        return (UserDetails) context.getAuthentication().getPrincipal();
+    }
+
+    private static String trimAndFormatDetail(final String formatString, final int expectedDetailLength) {
+        final String detail = StringUtils.defaultIfEmpty(formatString, StringUtils.EMPTY);
+        final String trimmedDetail = StringUtils.substring(detail, 0, expectedDetailLength);
```
Comment (schabdo) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java]: Here you do not use the multiplier of 2 !?!?!?

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java (hunk 2 above)
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java]: The username is = firstname & lastname for both parts you have a expected length e.g. 10. So the username is 20. 

I can switch the logic and you give the whole expected length of the username.
Then i have to divide by 2

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+    private static UserDetails loadUserByUsername(final String username) {
+        final UserDetailsService userDetailsService = SpringContextHelper.getBean(UserDetailsService.class);
+        try {
+            final UserDetails loadUserByUsername = userDetailsService.loadUserByUsername(username);
+            if (loadUserByUsername == null) {
+                throw new UsernameNotFoundException("User not found " + username);
+            }
+            return loadUserByUsername;
+        } catch (final UsernameNotFoundException e) { // NOSONAR
```
Comment (schabdo) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java]: At first glance empty catch but creating a dummy user below is not good for readability. What about this:

```
UserDetails  loadUserByUsername = null;

try {
    final UserDetailsService userDetailsService = SpringContextHelper.getBean(UserDetailsService.class);
    loadUserByUsername = userDetailsService.loadUserByUsername(username);
    if (loadUserByUsername == null) {
            throw new UsernameNotFoundException("User not found " + username);
    }
   
} catch (final UsernameNotFoundException e) {
     loadUserByUsername = new User(username, "", Collections.emptyList());
}

return loadUserByUsername;
```

Comment (schabdo) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java]: Maybe you can even remove `// NOSONAR` afterwards

Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java]: What about this?
````
        final UserDetailsService userDetailsService = SpringContextHelper.getBean(UserDetailsService.class);
        try {
            final UserDetails loadUserByUsername = userDetailsService.loadUserByUsername(username);
            if (loadUserByUsername == null) {
                throw new UsernameNotFoundException("User not found " + username);
            }
            return loadUserByUsername;
        } catch (final UsernameNotFoundException e) { 
            return new User(username, "", Collections.emptyList());
        }
```

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/DistributionBeanQuery.java [hunk 5]
```diff
@@ -62,7 +63,8 @@ public DistributionBeanQuery(final QueryDefinition definition, final Map<String,
             sort = new Sort(sortStates[0] ? Direction.ASC : Direction.DESC, (String) sortPropertyIds[0]);
             // Add sort
```
Comment (schabdo) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rollout/DistributionBeanQuery.java]: remove

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java (hunk 4 above)
Comment (schabdo) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java]: Even better. Go right ahead!

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java (hunk 2 above)
Comment (schabdo) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java]: I think that from the caller's point of view it is easier to say: "I have 20 characters of space left, please give me a suitable sized username". So it is easier to call, ready and understand. I do not care about the internals as long as there is no requirement that firstname and lastname have to be equal in size.

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+     *            the entity
+     * @return the formatted user name (max 12 characters) cannot be <null>
+     */
+    public static String loadAndFormatCurrentUsername() {
+        return loadAndFormatUsername(getCurrentUser().getUsername(), 12);
+    }
+
+    /**
+     * Load user details by the user name and format the user name. If the
+     * loaded {@link UserDetails} is not a instance of a {@link UserPrincipal},
```
Comment (melanieretter) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/UserDetailsFormatter.java]: **an** instance

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/SPDateTimeUtil.java [hunk 7]
```diff
@@ -80,14 +82,45 @@ public static ZoneId getTimeZoneId(final TimeZone tz) {
...
+     */
+    public static String formatCreatedAt(final BaseEntity baseEntity) {
+        if (baseEntity == null) {
+            return StringUtils.EMPTY;
+        }
+        return formatDate(baseEntity.getCreatedAt(), StringUtils.EMPTY);
+    }
+
+    /**
+     * Get formatted date 'last modefied at' by entity.
```
Comment (melanieretter) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/SPDateTimeUtil.java]: mod**i**fied

**Pull request merged**